test/subtypes.test.did

```diff
@@ -0,0 +1,164 @@
+assert blob "DIDL\02\6a\00\01\01\00\6a\00\00\00\01\00\01\01\00\01m"
+   == "(opt func \"aaaaa-aa\".m)" : (opt func () -> (func () -> (reserved))) "func () -> () <: func () -> (reserved)";
+assert blob "DIDL\02\6a\00\01\01\00\6a\00\00\00\01\00\01\01\00\01m"
+   == "(null)" : (opt func () -> (func () -> (empty))) "func () -> () </: func () -> (empty)";
+assert blob "DIDL\02\6a\00\01\01\00\6a\00\00\00\01\00\01\01\00\01m"
+   == "(null)" : (opt func () -> (func () -> (nat))) "func () -> () </: func () -> (nat)";
+assert blob "DIDL\02\6a\00\01\01\00\6a\00\00\00\01\00\01\01\00\01m"
+   == "(null)" : (opt func () -> (func () -> (null))) "func () -> () </: func () -> (null)";
+
+// optional func arguments
+assert blob "DIDL\03\6a\00\01\01\00\6a\01\02\00\00\6e\6f\01\00\01\01\00\01m"
+   == "(opt func \"aaaaa-aa\".m)" : (opt func () -> (func () -> ())) "func (opt empty) -> () <: func () -> ()";
+assert blob "DIDL\03\6a\00\01\01\00\6a\01\02\00\00\6e\7f\01\00\01\01\00\01m"
+   == "(opt func \"aaaaa-aa\".m)" : (opt func () -> (func () -> ())) "func (opt null) -> () <: func () -> ()";
+assert blob "DIDL\02\6a\00\01\01\00\6a\01\70\00\00\01\00\01\01\00\01m"
+   == "(opt func \"aaaaa-aa\".m)" : (opt func () -> (func () -> ())) "func (reserved) -> () <: func () -> ()";
+assert blob "DIDL\02\6a\00\01\01\00\6a\01\6f\00\00\01\00\01\01\00\01m"
+   == "(null)" : (opt func () -> (func () -> ())) "func (empty) -> () </: func () -> ()";
+assert blob "DIDL\02\6a\00\01\01\00\6a\01\7d\00\00\01\00\01\01\00\01m"
+   == "(null)" : (opt func () -> (func () -> ())) "func (nat) -> () </: func () -> ()";
+assert blob "DIDL\02\6a\00\01\01\00\6a\01\7f\00\00\01\00\01\01\00\01m"
+   == "(null)" : (opt func () -> (func () -> ())) "func (null) -> () </: func () -> ()";
+
+// variants
+assert blob "DIDL\02\6a\00\01\01\00\6b\00\01\00\01\01\00\01m"
+   == "(opt func \"aaaaa-aa\".m)" : (opt func () -> (variant {})) "variant {} <: variant {}";
+assert blob "DIDL\02\6a\00\01\01\00\6b\00\01\00\01\01\00\01m"
+   == "(opt func \"aaaaa-aa\".m)" : (opt func () -> (variant {0 : nat})) "variant {} <: variant {0 : nat}";
+assert blob "DIDL\02\6a\00\01\01\00\6b\01\00\7d\01\00\01\01\00\01m"
+   == "(opt func \"aaaaa-aa\".m)" : (opt func () -> (variant {0 : nat})) "variant {0 : nat} <: variant {0 : nat}";
+assert blob "DIDL\02\6a\00\01\01\00\6b\01\00\7e\01\00\01\01\00\01m"
+   == "(null)" : (opt func () -> (variant {0 : nat})) "variant {0 : bool} </: variant {0 : nat}";
+assert blob "DIDL\02\6a\00\01\01\00\6b\01\00\7e\01\00\01\01\00\01m"
+   == "(null)" : (opt func () -> (variant {1 : bool})) "variant {0 : bool} </: variant {1 : bool}";
+
+// infinite types
+
+assert blob "DIDL\02\6a\00\01\01\00\6c\01\00\01\01\00\01\01\00\01m"
+   == "(opt func \"aaaaa-aa\".m)" : (opt func () -> (EmptyRecord)) "(µ record) <: (µ record)";
+assert blob "DIDL\02\6a\00\01\01\00\6c\01\00\01\01\00\01\01\00\01m"
+   == "(null)" : (opt func () -> (empty)) "(µ record) </: empty";
```

Comment:
```suggestion
// assert blob "DIDL\02\6a\00\01\01\00\6c\01\00\01\01\00\01\01\00\01m"
//   == "(null)" : (opt func () -> (empty)) "(µ record) </: empty";
```

I will comment out this test for now. It's a bit implementation dependent. In Rust, we replace `µ record` with `empty` after parsing the type table, so we actually get `opt func` instead of `null`.

Reply:
Can you explain why? I don't quite follow.

Reply:
In the Haskell code I think I too replace `mu record` with empty, but only after subtype checking?

I believe Motoko gets this right as spec'ed.

If we want to allow implementations to differ here we should say so explicitly in the spec. But I don't know if that's a good idea.

Could you introduce in the rust code a `MuRecord` type that you can replace it with, which does the job of preventing your code to run into these vicious cycles, while still having the right subtypes relations?

Hmm, but `mu record <: 
mu (record opt)`, so you probably really can't throw away the structure before subtype checking.

Reply:
It’s safe, but it is not according to specification. If we go that route, it would be consistent to equate all other empty types as well (`empty`, `{empty, empty}` etc.), and it should be part of the spec.